fix(helm): sort release list with flags --reverse and --date

jan25 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/master/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
Bug fix and closes #5499

**Special notes for your reviewer**:
Apart from fixing the bug for `--date` and `--reverse` flags in `helm list` command, this PR also makes sure the behavior is as expected if both `--date` and `--reverse` flags are enabled
Precisely:
`helm list` (Default case - sort by name Asc)
`helm list -r` sorts by name Desc
`helm list -d` sorts by date Asc (recent updated release last)
`helm list -d -r` sorts by date Desc (recently updated release first)

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

jan25 committed:
fix sort list with options bug

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

jan25 committed:
fix test cases for sort list with options

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

bacongobbler commented:
Would you mind fixing up the failing test case?

```
--- FAIL: TestList_Sort (0.00s)
    list_test.go:97: 
        	Error Trace:	list_test.go:97
        	Error:      	Not equal: 
        	            	expected: "two"
        	            	actual  : "one"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-two
        	            	+one
        	Test:       	TestList_Sort
    list_test.go:99: 
        	Error Trace:	list_test.go:99
        	Error:      	Not equal: 
        	            	expected: "one"
        	            	actual  : "two"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-one
        	            	+two
        	Test:       	TestList_Sort
FAIL
```

jan25 committed:
handle default list sort order through enum order

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

hickeyma reviewed (commented):
Looks good, thanks. Unit tests are running ok and `helm list` is working for the  alphanumeric.  Should the date sort work across whole list or per release?

```console
$ helm list -d
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
fooxxxxxxxxxxxxx           	default  	1       	2019-04-02 14:08:03.560578811 +0100 IST	deployed	foo-0.1.0             
fooxxxxxxxxxxxxx-1554210671	default  	1       	2019-04-02 14:11:11.871312236 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
martychrt-1554198221       	default  	2       	2019-04-02 11:14:12.297504181 +0100 IST	deployed	martychrt-0.1.0       
foo-0.1.0-beta.1-1554210628	default  	1       	2019-04-02 14:10:28.122494225 +0100 IST	failed  	foo-0.1.0-beta.1-0.1.0

$ helm list -d -r
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
foo-0.1.0-beta.1-1554210628	default  	1       	2019-04-02 14:10:28.122494225 +0100 IST	failed  	foo-0.1.0-beta.1-0.1.0
martychrt-1554198221       	default  	2       	2019-04-02 11:14:12.297504181 +0100 IST	deployed	martychrt-0.1.0       
fooxxxxxxxxxxxxx-1554210671	default  	1       	2019-04-02 14:11:11.871312236 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
fooxxxxxxxxxxxxx           	default  	1       	2019-04-02 14:08:03.560578811 +0100 IST	deployed	foo-0.1.0  
```

jan25 commented:
@hickeyma Hi, Thanks for the review!

Sorry, I couldn't understand the snippet in your comment. Shouldn't it be as below?
```
$ helm list -d
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
martychrt-1554198221       	default  	2       	2019-04-02 11:14:12.297504181 +0100 IST	deployed	martychrt-0.1.0       
fooxxxxxxxxxxxxx           	default  	1       	2019-04-02 14:08:03.560578811 +0100 IST	deployed	foo-0.1.0             
foo-0.1.0-beta.1-1554210628	default  	1       	2019-04-02 14:10:28.122494225 +0100 IST	failed  	foo-0.1.0-beta.1-0.1.0
fooxxxxxxxxxxxxx-1554210671	default  	1       	2019-04-02 14:11:11.871312236 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0

$ helm list -d -r
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
fooxxxxxxxxxxxxx-1554210671	default  	1       	2019-04-02 14:11:11.871312236 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
foo-0.1.0-beta.1-1554210628	default  	1       	2019-04-02 14:10:28.122494225 +0100 IST	failed  	foo-0.1.0-beta.1-0.1.0
fooxxxxxxxxxxxxx           	default  	1       	2019-04-02 14:08:03.560578811 +0100 IST	deployed	foo-0.1.0
martychrt-1554198221       	default  	2       	2019-04-02 11:14:12.297504181 +0100 IST	deployed	martychrt-0.1.0       
```
Also, didn't quite understand your question. I believe we have one row per release, right? unlike `helm history`

bacongobbler commented:
@jan25's output is correct. The date sort should only work against the displayed release versions. When the user inputs `helm list --date`, the intent is to sort the releases by the time each release was last updated.

bacongobbler commented on pkg/action/list.go:
```diff
@@ -87,12 +87,14 @@ const ListAll = ListDeployed | ListUninstalled | ListUninstalling | ListPendingI
-	ByDate Sorter = iota
 	// ByNameAsc sorts by ascending lexicographic order
-	ByNameAsc
+	ByNameAsc Sorter = iota
```
is this still in use? If not, is it safe to remove this? I think the `default` case below captures this enum so there's no need to use this if I'm mistaken.

jan25 replied:
Nice catch! It wasn't directly used, but only in test cases to enforce default sort order. I fixed the test cases to remove its usage and adjusted the enum fields.

jan25 committed:
Remove redundant ByNameAsc enum value. Handled by default sorting

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

bacongobbler commented:
@adamreese would you mind taking a look?

hickeyma commented:
>  The date sort should only work against the displayed release versions. When the user inputs helm list --date, the intent is to sort the releases by the time each release was last updated.

@bacongobbler @jan25 Sorry if I didn't clarify the comment above correctly. I agree with your definition on how it work. However when I test from the CLI with releases previously deployed, this is the output I am getting which seems incorrect to me:
```console
$ helm list -d
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
fooxxxxxxxxxxxxx           	default  	1       	2019-04-02 14:08:03.560578811 +0100 IST	deployed	foo-0.1.0             
fooxxxxxxxxxxxxx-1554210671	default  	1       	2019-04-02 14:11:11.871312236 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
martychrt-1554198221       	default  	2       	2019-04-02 11:14:12.297504181 +0100 IST	deployed	martychrt-0.1.0       
foo-0.1.0-beta.1-1554210628	default  	1       	2019-04-02 14:10:28.122494225 +0100 IST	failed  	foo-0.1.0-beta.1-0.1.0

$ helm list -d -r
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
foo-0.1.0-beta.1-1554210628	default  	1       	2019-04-02 14:10:28.122494225 +0100 IST	failed  	foo-0.1.0-beta.1-0.1.0
martychrt-1554198221       	default  	2       	2019-04-02 11:14:12.297504181 +0100 IST	deployed	martychrt-0.1.0       
fooxxxxxxxxxxxxx-1554210671	default  	1       	2019-04-02 14:11:11.871312236 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
fooxxxxxxxxxxxxx           	default  	1       	2019-04-02 14:08:03.560578811 +0100 IST	deployed	foo-0.1.0  
```

jan25 commented:
@hickeyma are you using helm cli which is built with this PR? I tested it on my local machine, it looked ok for than 2 releases. Although, the test cases have just 2 releases. From your example, I suspect you might be running helm v3 missing changes from this PR, could you confirm?

hickeyma commented:
@jan25 I have picked up your changes. The reverse works for alphanumeric:
```console
$ helm list
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
foo-0.1.0-beta.1-1554210628	default  	1       	2019-04-02 14:10:28.122494225 +0100 IST	failed  	foo-0.1.0-beta.1-0.1.0
fooxxxxxxxxxxxxx           	default  	1       	2019-04-02 14:08:03.560578811 +0100 IST	deployed	foo-0.1.0             
fooxxxxxxxxxxxxx-1554210671	default  	1       	2019-04-02 14:11:11.871312236 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
martychrt-1554198221       	default  	2       	2019-04-02 11:14:12.297504181 +0100 IST	deployed	martychrt-0.1.0       
usr1@helm-v3:~/test/helm-charts$ helm list -r
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
martychrt-1554198221       	default  	2       	2019-04-02 11:14:12.297504181 +0100 IST	deployed	martychrt-0.1.0       
fooxxxxxxxxxxxxx-1554210671	default  	1       	2019-04-02 14:11:11.871312236 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
fooxxxxxxxxxxxxx           	default  	1       	2019-04-02 14:08:03.560578811 +0100 IST	deployed	foo-0.1.0             
foo-0.1.0-beta.1-1554210628	default  	1       	2019-04-02 14:10:28.122494225 +0100 IST	failed  	foo-0.1.0-beta.1-0.1.0
```

hickeyma commented:
@jan25 As further testing, could you try and add a few more releases to your current list with different update times?

jan25 commented:
@hickeyma I'll improve the test cases to have more releases

jan25 committed:
fix docs for helm list

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

jan25 committed:
add more releases to list tests

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

jan25 committed:
add superseded to list filter mask

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

jan25 committed:
remove unnecessary setting of list.All flag

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

jan25 committed:
fix test cases for list.AllNamespaces

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

jan25 committed:
update list tests expected output files

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

jan25 commented:
@hickeyma I added two more releases to test cases and updated the expected output files for tests. Meanwhile i found issues in `list.go` such as `--all` and `-- superseded` weren't working as expected, so i fixed them too in recent commits. It was a good idea to improve and have robust test cases.

@bacongobbler Could you guys please do a review once again considering the recent changes i have made?

hickeyma reviewed (changes requested):
@jan25 Tested this again from the CLI.  Date sort seems to still have issues .It works ok for 2 releases but once you increase the releases, the sort seems to go awry. Here is output from my tests:
```console
$ helm ls
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed	foo-0.1.0             
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed	foo-0.1.0             
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
mychart-1554717818         	default  	2       	2019-04-08 11:25:09.648980346 +0100 IST	deployed	mychart-0.1.0         
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	deployed	nxtxhart-0.1.0        

$ helm ls -r
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	deployed	nxtxhart-0.1.0        
mychart-1554717818         	default  	2       	2019-04-08 11:25:09.648980346 +0100 IST	deployed	mychart-0.1.0         
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed	foo-0.1.0             
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed	foo-0.1.0             

$ helm ls -d
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
mychart-1554717818         	default  	2       	2019-04-08 11:25:09.648980346 +0100 IST	deployed	mychart-0.1.0         
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	deployed	nxtxhart-0.1.0        
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed	foo-0.1.0             
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed	foo-0.1.0             

$ helm ls -d -r
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed	foo-0.1.0             
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed	foo-0.1.0             
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	deployed	nxtxhart-0.1.0        
mychart-1554717818         	default  	2       	2019-04-08 11:25:09.648980346 +0100 IST	deployed	mychart-0.1.0        
 
$ helm ls --all
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS    	CHART                 
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed  	foo-0.1.0             
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed  	foo-0.1.0             
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed  	fooxxxxxxxxxxxxx-0.1.0
mychart-1554717818         	default  	1       	2019-04-08 11:03:38.789623647 +0100 IST	superseded	mychart-0.1.0         
mychart-1554717818         	default  	2       	2019-04-08 11:25:09.648980346 +0100 IST	deployed  	mychart-0.1.0         
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	

$ helm ls --superseded
NAME              	NAMESPACE	REVISION	UPDATED                                	STATUS    	CHART        
mychart-1554717818	default  	1       	2019-04-08 11:03:38.789623647 +0100 IST	superseded	mychart-0.1.0

jan25 committed:
fix ByDate sorter to use Time.Unix()

Signed-off-by: Abhilash Gnan <abhilashgnan@gmail.com>

jan25 commented:
@hickeyma I couldn't reproduce on my machine at first, but using the output you provided i could spot a bug in ByDate sorter. Thanks for checking!

hickeyma reviewed (approved):
LGTM now, thanks for all the updates. Output as follows on CLI test:

```console
$ helm ls
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed	foo-0.1.0             
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed	foo-0.1.0             
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
mychart-1554717818         	default  	3       	2019-04-09 14:54:16.471790642 +0100 IST	deployed	mychart-0.1.0         
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	deployed	nxtxhart-0.1.0        
xack-chart-1554817980      	default  	1       	2019-04-09 14:53:00.986993461 +0100 IST	deployed	xack-chart-0.1.0      

$ helm ls -r
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
xack-chart-1554817980      	default  	1       	2019-04-09 14:53:00.986993461 +0100 IST	deployed	xack-chart-0.1.0      
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	deployed	nxtxhart-0.1.0        
mychart-1554717818         	default  	3       	2019-04-09 14:54:16.471790642 +0100 IST	deployed	mychart-0.1.0         
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed	foo-0.1.0             
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed	foo-0.1.0             

$ helm ls -d
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	deployed	nxtxhart-0.1.0        
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed	foo-0.1.0             
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed	foo-0.1.0             
xack-chart-1554817980      	default  	1       	2019-04-09 14:53:00.986993461 +0100 IST	deployed	xack-chart-0.1.0      
mychart-1554717818         	default  	3       	2019-04-09 14:54:16.471790642 +0100 IST	deployed	mychart-0.1.0         

$ helm ls -d -r
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART                 
mychart-1554717818         	default  	3       	2019-04-09 14:54:16.471790642 +0100 IST	deployed	mychart-0.1.0         
xack-chart-1554817980      	default  	1       	2019-04-09 14:53:00.986993461 +0100 IST	deployed	xack-chart-0.1.0      
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed	foo-0.1.0             
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed	foo-0.1.0             
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	deployed	nxtxhart-0.1.0        
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed	fooxxxxxxxxxxxxx-0.1.0
(reverse-i-search)`all': helm inst^Cl --generate-name xack-chart/

$ helm ls --all
NAME                       	NAMESPACE	REVISION	UPDATED                                	STATUS    	CHART                 
foo                        	default  	1       	2019-04-08 11:11:38.009118174 +0100 IST	deployed  	foo-0.1.0             
foo-1554718257             	default  	1       	2019-04-08 11:10:57.93880877 +0100 IST 	deployed  	foo-0.1.0             
fooxxxxxxxxxxxxx-1554717509	default  	1       	2019-04-08 10:58:29.428715141 +0100 IST	deployed  	fooxxxxxxxxxxxxx-0.1.0
mychart-1554717818         	default  	1       	2019-04-08 11:03:38.789623647 +0100 IST	superseded	mychart-0.1.0         
mychart-1554717818         	default  	2       	2019-04-08 11:25:09.648980346 +0100 IST	superseded	mychart-0.1.0         
mychart-1554717818         	default  	3       	2019-04-09 14:54:16.471790642 +0100 IST	deployed  	mychart-0.1.0         
nxtxhart-1554717909        	default  	1       	2019-04-08 11:05:09.973096741 +0100 IST	deployed  	nxtxhart-0.1.0        
xack-chart-1554817980      	default  	1       	2019-04-09 14:53:00.986993461 +0100 IST	deployed  	xack-chart-0.1.0      

$ helm ls --superseded
NAME              	NAMESPACE	REVISION	UPDATED                                	STATUS    	CHART        
mychart-1554717818	default  	1       	2019-04-08 11:03:38.789623647 +0100 IST	superseded	mychart-0.1.0
mychart-1554717818	default  	2       	2019-04-08 11:25:09.648980346 +0100 IST	superseded	mychart-0.1.0
```

jan25 commented:
Is it a good idea to add `v3` tag to this PR? How can i do that? Perhaps it can be done by maintainers only?

hickeyma commented:
@adamreese Do you mind taking a look when you get a chance?

bacongobbler commented on cmd/helm/list.go:
```diff
@@ -66,7 +66,6 @@ func newListCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
 			if client.AllNamespaces {
 				client.SetConfiguration(newActionConfig(true))
 			}
-			client.All = client.Limit == -1
```
it looks like this got removed but nothing was added back to restore this functionality. Was this intentional, and if so is there a way we can add this back in?

jan25 replied:
Hi @bacongobbler !
I almost forgot why i removed this :)

This was intentionally done in this commit https://github.com/helm/helm/pull/5539/commits/f4c2b02cef95aecad6d691b812e4508fefb7ac6b. When i was fixing actual issue with the flags for #5499 i also found issues with `--all` and `--superseded` flags, so as part of fixing them i removed this line.
Also, the test cases were buggy for these ^ two flags so CI didn't complain so far about them, which is why i had to adjust/improve the test case logic and also adjust release input and output testdata.

Thanks for the review!

bacongobbler reviewed (changes requested):
Thank you for taking the time to fix this, @jan25. Sorry for the late response here. I have one more question about something that was changed in this PR. Once that comment's been addressed I think we're good to go!

bacongobbler reviewed (approved):
LGTM!
